Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp

```diff
@@ -0,0 +1,151 @@
+    ASSERT(targetRenderer.hasLayer());
+
+    auto& context = paintInfo.context();
+    GraphicsContextStateSaver stateSaver(context);
+
+    auto objectBoundingBox = targetRenderer.objectBoundingBox();
+    auto boundingBoxTopLeftCorner = flooredLayoutPoint(objectBoundingBox.minXMinYCorner());
+    auto coordinateSystemOriginTranslation = adjustedPaintOffset - boundingBoxTopLeftCorner;
+    if (!coordinateSystemOriginTranslation.isZero())
+        context.translate(coordinateSystemOriginTranslation);
+
+    AffineTransform contentTransform;
+    auto& maskElement = this->maskElement();
+    if (maskElement.maskContentUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX) {
+        contentTransform.translate(objectBoundingBox.x(), objectBoundingBox.y());
+        contentTransform.scale(objectBoundingBox.width(), objectBoundingBox.height());
+    }
+
+    auto repaintBoundingBox = targetRenderer.repaintRectInLocalCoordinates();
+    auto absoluteTransform = context.getCTM(GraphicsContext::DefinitelyIncludeDeviceScale);
+
+    auto maskColorSpace = DestinationColorSpace::SRGB();
+    auto drawColorSpace = DestinationColorSpace::SRGB();
+
+    const auto& svgStyle = style().svgStyle();
+#if ENABLE(DESTINATION_COLOR_SPACE_LINEAR_SRGB)
+    if (svgStyle.colorInterpolation() == ColorInterpolation::LinearRGB) {
+#if USE(CG)
+        maskColorSpace = DestinationColorSpace::LinearSRGB();
+#endif
+        drawColorSpace = DestinationColorSpace::LinearSRGB();
+    }
+#endif
+
+    auto maskImage = SVGRenderingContext::createImageBuffer(repaintBoundingBox, absoluteTransform, maskColorSpace, context.renderingMode());
+    if (!maskImage)
+        return;
+
+    context.setCompositeOperation(CompositeOperator::DestinationIn);
+    context.beginTransparencyLayer(1);
+
+    auto& maskImageContext = maskImage->context();
+    layer()->paintSVGResourceLayer(maskImageContext, stateSaver, contentTransform);
+
+#if !USE(CG)
+    maskImage->transformToColorSpace(drawColorSpace);
+#else
+    UNUSED_PARAM(drawColorSpace);
+#endif
+
+    if (svgStyle.maskType() == MaskType::Luminance)
+        maskImage->convertToLuminanceMask();
+
+    context.setCompositeOperation(CompositeOperator::SourceOver);
+
+    // The mask image has been created in the absolute coordinate space, as the image should not be scaled.
+    // So the actual masking process has to be done in the absolute coordinate space as well.
+    FloatRect absoluteTargetRect = enclosingIntRect(absoluteTransform.mapRect(repaintBoundingBox));
+    context.concatCTM(absoluteTransform.inverse().value_or(AffineTransform()));
+    context.drawConsumingImageBuffer(WTFMove(maskImage), absoluteTargetRect);
+    context.endTransparencyLayer();
+}
+
+FloatRect RenderSVGResourceMasker::resourceBoundingBox(const RenderObject& object, RepaintRectCalculation)
+{
+    FloatRect targetBoundingBox = object.objectBoundingBox();
+
+    // Resource was not layouted yet. Give back the boundingBox of the object.
```

Comment:
Not sure how relevant that is in LBSE.

Reply:
Good point ,should it be an ASSERT instead or just remove the check?

Reply:
Sure, done!